src/Microsoft.PowerShell.ConsoleHost/host/msh/UpdatesNotification.cs

```diff
@@ -0,0 +1,326 @@
+                        }
+                    }
+
+                    // Finally, create the `todayDoneFilePath` file as an indicator that a successful update check has finished today.
+                    new FileStream(todayDoneFilePath, FileMode.CreateNew, FileAccess.Write, FileShare.None).Close();
+                }
+            }
+            catch (Exception)
+            {
+                // There are 2 possible reason for the exception:
+                // 1. An update check initiated from another `pwsh` process is in progress.
+                //    It's OK to just return and let that update check to finish the work.
+                // 2. The update check failed (ex. internet connectivity issue, GitHub service failure).
+                //    It's OK to just return and let another `pwsh` do the check at later time.
+            }
+        }
+
+        private static bool TryParseUpdateFile(
+            out string updateFilePath,
+            out SemanticVersion lastUpdateVersion,
+            out DateTime lastUpdateDate)
+        {
+            updateFilePath = null;
+            lastUpdateVersion = null;
+            lastUpdateDate = default;
+
+            var files = Directory.EnumerateFiles(s_cacheDirectory, UpdateFileNamePattern, s_enumOptions);
+            var enumerator = files.GetEnumerator();
+
+            if (!enumerator.MoveNext())
+            {
+                // No file was found that matches the pattern, but it's OK that an update file doesn't exist.
+                // This could happen when there is no new updates yet.
+                return true;
+            }
+
+            updateFilePath = enumerator.Current;
+            if (enumerator.MoveNext())
+            {
+                // More than 1 files were found that match the pattern. This is a corrupted state.
+                // Theoretically, there should be only one update file at any point of time.
+                updateFilePath = null;
+                return false;
+            }
+
+            // OK, only found one update file, which is expected.
+            // Now let's parse the file name.
+            string updateFileName = Path.GetFileName(updateFilePath);
+            int dateStartIndex = updateFileName.LastIndexOf('_') + 1;
+
+            bool success = DateTime.TryParse(
+                updateFileName.AsSpan().Slice(dateStartIndex),
+                CultureInfo.InvariantCulture,
+                DateTimeStyles.AssumeLocal,
+                out lastUpdateDate);
+
+            if (success)
+            {
+                int versionStartIndex = updateFileName.IndexOf('_') + 2;
+                int versionLength = dateStartIndex - versionStartIndex - 1;
+                string versionString = updateFileName.Substring(versionStartIndex, versionLength);
+                success = SemanticVersion.TryParse(versionString, out lastUpdateVersion);
+            }
+
+            if (!success)
```

Comment:
This all seems a bit harder to follow than:

```csharp
if (DateTime.TryParse(
        updateFileName.AsSpan().Slice(dateStartIndex),
        CultureInfo.InvariantCulture,
        DateTimeStyles.AssumeLocal,
        out lastUpdateDate))
{
    int versionStartIndex = updateFileName.IndexOf('_') + 2;
    int versionLength = dateStartIndex - versionStartIndex - 1;
    string versionString = updateFileName.Substring(versionStartIndex, versionLength);

    if (SemanticVersion.TryParse(versionString, out lastUpdateVersion)
    {
        return true;
    }
}

return false;
```

Or (of course):

```csharp
if (!DateTime.TryParse(
        updateFileName.AsSpan().Slice(dateStartIndex),
        CultureInfo.InvariantCulture,
        DateTimeStyles.AssumeLocal,
        out lastUpdateDate))
{
    return false;
}

int versionStartIndex = updateFileName.IndexOf('_') + 2;
int versionLength = dateStartIndex - versionStartIndex - 1;
string versionString = updateFileName.Substring(versionStartIndex, versionLength);

return SemanticVersion.TryParse(versionString, out lastUpdateVersion);
```

Reply:
When `SemanticVersion.TryParse` fails (return `false`), both `lastUpdateDate` and `updateFilePath` need to be set to `default<T>`.
I still think the current organization of the code is better.

Reply:
The part I don't like about it is having to follow the state of `success`. We have other code in PowerShell that does that and perhaps it started out simple enough, but it gets out of hand quickly. I found just this usage more effort to follow because I had to think about the mutating state of `success`.

The format of the code is just style and I'm not attached to it, but I really think we should avoid rolling, mutating booleans.

Reply:
OK will change the code to return early and see how it looks.

Reply:
Updated. Please take another look. Thanks!